### PR TITLE
Update docs to adjust the updated demo

### PR DESCRIPTION
**Pull request by alonre24**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,133 +1,112 @@
...
+To run the demo app and load the data, run the following:
 ```
-$ git clone https://github.com/RedisAI/FraudDetectionDemo.git
-$ cd FraudDetectionDemo
 # If you don't have it already, install https://git-lfs.github.com/ (On OSX: brew install git-lfs)
 $ git lfs install && git lfs fetch && git lfs checkout
+
+# Clone the repository
+$ git clone https://github.com/RedisAI/FraudDetectionDemo.git
+$ cd FraudDetectionDemo
```
Comment (DvirDukhan) [inline README.md:19]: this is done prior to git lfs

File: README.md [hunk 2]
```diff
@@ -1,133 +1,112 @@
...
 $ docker-compose up --force-recreate --build
 ```
 
-The `app` container will [load](https://github.com/RedisAI/FraudDetectionDemo/blob/master/app/app.py)
-- The [two tensorflow models](https://github.com/RedisAI/FraudDetectionDemo/tree/master/app/models)
-- The [Torch script](https://github.com/RedisAI/FraudDetectionDemo/blob/master/app/script.torch)
-- The RedisGears functions
+## Architecture
+### RedisAI
+Redis instance is launched with RedisAI module loaded (by running `redislabs/redisai:latest` docker). Recall that RedisAI allows you to store and execute AI/ML pre-trained models in Redis.
```
Comment (DvirDukhan) [inline README.md:31]:
```suggestion
Redis instance is launched with RedisAI module loaded (by running `redislabs/redisai:latest` docker). RedisAI allows you to store and execute AI/ML pre-trained models in Redis.
```

File: README.md [hunk 3]
```diff
@@ -1,133 +1,112 @@
...
+
+### Data Loader
```
Comment (DvirDukhan) [inline README.md:33]: link to the file

File: README.md [hunk 4]
```diff
@@ -1,133 +1,112 @@
...
+Historical data of credit card transactions is loaded into Redis from `data/creditcard.csv` file, using redis-py (Redis' official python client). Every transaction is stored in Redis as a hash with the following structure:
+- The key is `<time_stamp>_<index>{tag}`, where `<index>` is used to differentiate between keys which correspond to transactions that have occurred in the same timestamp, and `{tag}` is used to ensure that all hashes are stored in the same shard (when using clustered environment). 
+- The hash value is a dictionary that contains 29 numeric features ("v0", "v1", ... "v28") that represent the transaction, and another field ("amount") specifying the amount of money to transfer in the transaction.
+
+In addition, all the hashes' keys are kept in a sorted set stored in Redis whose key is `refernces{tag}`, and every element in it is of the form: `{hash_key_name: timestamp}`. This sorted set keeps track of the keynames of these hashes sorted by time.
```
Comment (DvirDukhan) [inline README.md:38]: link to line

File: README.md [hunk 5]
```diff
@@ -1,133 +1,112 @@
...
-- The [two tensorflow models](https://github.com/RedisAI/FraudDetectionDemo/tree/master/app/models)
-- The [Torch script](https://github.com/RedisAI/FraudDetectionDemo/blob/master/app/script.torch)
-- The RedisGears functions
+## Architecture
+### RedisAI
+Redis instance is launched with RedisAI module loaded (by running `redislabs/redisai:latest` docker). Recall that RedisAI allows you to store and execute AI/ML pre-trained models in Redis.
+
+### Data Loader
+Historical data of credit card transactions is loaded into Redis from `data/creditcard.csv` file, using redis-py (Redis' official python client). Every transaction is stored in Redis as a hash with the following structure:
+- The key is `<time_stamp>_<index>{tag}`, where `<index>` is used to differentiate between keys which correspond to transactions that have occurred in the same timestamp, and `{tag}` is used to ensure that all hashes are stored in the same shard (when using clustered environment).
```
Comment (DvirDukhan) [inline README.md:35]: link to line

File: README.md [hunk 6]
```diff
@@ -1,133 +1,112 @@
...
+- The hash value is a dictionary that contains 29 numeric features ("v0", "v1", ... "v28") that represent the transaction, and another field ("amount") specifying the amount of money to transfer in the transaction.
+
+In addition, all the hashes' keys are kept in a sorted set stored in Redis whose key is `refernces{tag}`, and every element in it is of the form: `{hash_key_name: timestamp}`. This sorted set keeps track of the keynames of these hashes sorted by time.
 
-The `dataloader` container will [load](https://github.com/RedisAI/FraudDetectionDemo/blob/master/dataloader/load.py)
-- 1000 [raw reference data points](https://github.com/RedisAI/FraudDetectionDemo/blob/master/dataloader/load.py#L29)
-- 1000 [tensor](https://github.com/RedisAI/FraudDetectionDemo/blob/master/dataloader/load.py#L32) representations of these raw data points
-- 1 [sorted set](https://github.com/RedisAI/FraudDetectionDemo/blob/master/dataloader/load.py#L35)
+### App
```
Comment (DvirDukhan) [inline README.md:40]: link to file

File: README.md [hunk 7]
```diff
@@ -1,133 +1,112 @@
...
+The app is a basic component that loads the fraud detection pre-trained ML model into RedisAI, along with a [TorchScript](https://oss.redis.com/redisai/intro/#scripting) which is used for pre-processing of the data. This model was built using Tensorflow, which is one of the three ML frameworks that RedisAI supports as backends.
```
Comment (DvirDukhan) [inline README.md:41]: lines where this action is happen

File: README.md [hunk 8]
```diff
@@ -1,133 +1,112 @@
...
 
-### Explore loaded reference data
-This demo bundles a [RedisInsight](https://redislabs.com/redisinsight/) container which is an intuitive GUI for Redis. Open a browser and point it at https://localhost:8001 and select the preloaded redis connection.
+**Multiple devices:** RedisAI allows executing operation in parallel on different logical devices. Hence, the fraud-detection model is loaded twice under two different key names: `fraud_detection_model{tag}_CPU` and `fraud_detection_model{tag}_CPU:1`, which are associated to `CPU` and `CPU:1` devices respectively. Side note - If your machine has GPU with nvidia CUDA support, it is a good practice to load models that can run in parallel and associate each one with a different device (for example: `fraud_detection_model{tag}_GPU:0` and `fraud_detection_model{tag}_GPU:1`), for gaining better utilization of GPU resources.
```
Comment (DvirDukhan) [inline README.md:43]: can you re write to something like:
1. `fraud_detection_model{tag}_CPU` will be saved on `CPU`
2. `fraud_detection_model{tag}_CPU:1` will be saved on `CPU:1`

also, for using GPU:0, GPU:1 you will need a multiple GPU machine so the last sentence is not accurate

File: README.md [hunk 9]
```diff
@@ -1,133 +1,112 @@
...
-"Value at:0x7f915600dc10 refcount:1 encoding:raw serializedlength:146 lru:16636612 lru_seconds_idle:69"
->> hset 478_2 Amount 123
-(integer) 0
->> debug object 478_2_tensor
-"Value at:0x7f915600dc20 refcount:1 encoding:raw serializedlength:146 lru:16636690 lru_seconds_idle:2"
-```
-![Demo Flow 1 CLI](./demo_flow1_cli.png "Demo Flow 1 CLI")
+## Flow
+When a new transaction happens, it needs to be evaluated if it's fraudulent or not. The ML/DL models take two inputs for this, the relevant reference data and the new transaction details.
+First, the relevant keys with respect to the new transaction time and a certain time interval are obtained via range query over the `refernces` sorted set. The hash values of those keys will constitute the relevant reference data. Next, RedisAI DAG (directed acyclic graph) object will is used to orchestrate the end-to-end prediction flow, that will be triggered upon calling the [AI.DAGEXECUTE command](https://oss.redis.com/redisai/commands/#aidagexecute). The DAG's operation are the following:
```
Comment (DvirDukhan) [inline README.md:72]: break this long sentence into bullets please

File: README.md [hunk 10]
```diff
@@ -1,133 +1,112 @@
...
+1. Loading the numeric representation of the new transaction as a tensor with [`AI.TENSROSET` command](https://oss.redis.com/redisai/commands/#aitensorset).
```
Comment (DvirDukhan) [inline README.md:73]: link to the actual code

File: README.md [hunk 11]
```diff
@@ -1,133 +1,112 @@
...
+2. Running the `hashes_to_tensor` pre-processing function to create the reference data tensor. This function receives the list of relevant keys in Redis as input, and construct a reference data tensor based on these values, that are fetched directly from Redis (using [Redis command support for TorchScript](https://oss.redis.com/redisai/commands/#redis-commands-support)).
```
Comment (DvirDukhan) [inline README.md:74]: link to the function

File: README.md [hunk 12]
```diff
@@ -1,133 +1,112 @@
...
+3. Running the two fraud-detection models over the `transaction` and `refernces` inputs in parallel (recall that each model was associated with a different logical device).
```
Comment (DvirDukhan) [inline README.md:75]: recall == point to where you wrote it

File: README.md [hunk 13]
```diff
@@ -1,133 +1,112 @@
...
+4. Running the `post_processing` function that receives both models' predictions as inputs, and outputs the average score.
```
Comment (DvirDukhan) [inline README.md:76]: link to the function

File: README.md (hunk 1 above)
Comment (alonre24) [reply README.md:19]: Are you sure? I cloned before I had git-lfs installed, and only when I removed and re-cloned the repo it worked for me...

File: README.md (hunk 8 above)
Comment (alonre24) [reply README.md:43]: OK, I will rephrase

File: README.md [hunk 14]
```diff
@@ -1,133 +1,114 @@
...
-
-![Gears<->Redis data gathering](./flow3.png "Gears<->Redis data gathering")
-
-From this list it extracts a list of tensors from the keyspace and sends it to a [Torch script](https://github.com/RedisAI/FraudDetectionDemo/blob/master/app/script.torch). This torch script creates a
-new tensor with the shape `(1, 256)`, by concatenating the tensors and either pad the remaining space or trimming it.
-The new tensor is the reference data for the models which expect a reference data with shape of `(1, 256)` and transaction details tensor with the shape of `(1, 30)`. Once the inferencing of the models are done, the function uses `numpy` to aggregate the results and save them to a tensor
-with shape `(1, 2)` that contains the probability for the transaction to be a fraud.
-
-![Gears<->AI execution](./flow4.png "Gears<->AI execution")
+This demo showcases the advantage of **data locality** during transaction scoring, by using [RedisAI](https://oss.redislabs.com/redisai/) module.
```
Comment (esandoval30) [inline README.md:8]: This demo illustrates the key advantage of **data locality** using Redis AI (link to the module). Generating a prediction from a real-time AI/L model is a multi-step process. The steps typically include: receiving a prediction (inference) request, retrieving feature data (needed by the mode)l and running the inference request (feeding features into model). In general, these steps run in multiple processes/machines. In contrast, using Redis and the RedisAI model allows you store feature data and run the AI/Model in Redis! The feature data is within easy reach of the model!

File: README.md [hunk 15]
```diff
@@ -1,133 +1,114 @@
...
+It simulates a fraud-detection app which is based on ML model prediction in real-time. When a new transaction is executed, it triggers a flow whose output is the probability that this transaction is fraudulent. The prediction is based on both the given transaction and some reference data stored in Redis, while *the entire flow* is done within Redis as well.
```
Comment (esandoval30) [inline README.md:9]: This demo simulates a fraud-detection app, which relies on a ML model returning the probability that a given transaction is fraudulent. The model uses transaction data and reference (feature) data previously stored in Redis for the customers. The entire flow is executed within redis

File: README.md [hunk 16]
```diff
@@ -1,133 +1,114 @@
...
-The new tensor is the reference data for the models which expect a reference data with shape of `(1, 256)` and transaction details tensor with the shape of `(1, 30)`. Once the inferencing of the models are done, the function uses `numpy` to aggregate the results and save them to a tensor
-with shape `(1, 2)` that contains the probability for the transaction to be a fraud.
-
-![Gears<->AI execution](./flow4.png "Gears<->AI execution")
+This demo showcases the advantage of **data locality** during transaction scoring, by using [RedisAI](https://oss.redislabs.com/redisai/) module.
+It simulates a fraud-detection app which is based on ML model prediction in real-time. When a new transaction is executed, it triggers a flow whose output is the probability that this transaction is fraudulent. The prediction is based on both the given transaction and some reference data stored in Redis, while *the entire flow* is done within Redis as well.   
 
 ## Running the Demo
-To run the demo:
+To run the demo app and load the data, run the following:
```
Comment (esandoval30) [inline README.md:12]: run the following commands

File: README.md (hunk 1 above)
Comment (esandoval30) [reply README.md:19]: I think you need lfs first

File: README.md [hunk 17]
```diff
@@ -1,133 +1,114 @@
...
 $ docker-compose up --force-recreate --build
 ```
 
-The `app` container will [load](https://github.com/RedisAI/FraudDetectionDemo/blob/master/app/app.py)
-- The [two tensorflow models](https://github.com/RedisAI/FraudDetectionDemo/tree/master/app/models)
-- The [Torch script](https://github.com/RedisAI/FraudDetectionDemo/blob/master/app/script.torch)
-- The RedisGears functions
+## Architecture
+### RedisAI
+Redis instance is launched with RedisAI module loaded (by running `redislabs/redisai:latest` docker). RedisAI allows you to store and execute AI/ML pre-trained models in Redis.
```
Comment (esandoval30) [inline README.md:31]: ....execute AI/ML models in Redis. The models can be written in Pytorch, Tensorflow or any other ML framework (XGBoot, scikit-learn) allowing models to be exported to ONNX

File: README.md [hunk 18]
```diff
@@ -1,133 +1,114 @@
...
+
+### [Data Loader](https://github.com/RedisAI/FraudDetectionDemo/blob/master/dataloader/load.py)
+Historical data of credit card transactions is loaded into Redis from [`data/creditcard.csv` file](https://media.githubusercontent.com/media/RedisAI/FraudDetectionDemo/master/dataloader/data/creditcard.csv), using redis-py (Redis' official python client). Every transaction is [stored in Redis](https://github.com/RedisAI/FraudDetectionDemo/blob/master/dataloader/load.py#L31) as a hash with the following structure:
```
Comment (esandoval30) [inline README.md:34]: What kind of historical data is this? I presume some feature data?

**Review by esandoval30 — APPROVED** [review]

File: README.md (hunk 18 above)
Comment (alonre24) [reply README.md:34]: I guess so... The 29 numeric values are in the range of `(0, 1)` and have no clear meaning... how would you call it?

File: README.md (hunk 8 above)
Comment (esandoval30) [reply README.md]: minor rephrase **Multiple devices** RedisAI allows parallel execution of models on different logical devices....In this demo, the fraud detection model is loaded on 2 logical devices

File: README.md (hunk 18 above)
Comment (esandoval30) [reply README.md]: model features! 

I imagine that These numbers are the result of some feature engineering process that takes raw features as inputs!

Comment (alonre24) [reply README.md]: I rephrased to "Historical feature-data". ok?

Comment (alonre24) [reply README.md]: or maybe "Historical data of credit card transactions which has been processed into features is loaded..."
